pkg/cmd/pulumi/state_upgrade.go

```diff
@@ -101,5 +110,89 @@ func (cmd *stateUpgradeCmd) Run(ctx context.Context) error {
 		return nil
 	}
 
-	return lb.Upgrade(ctx, nil /*opts*/)
+	var opts filestate.UpgradeOptions
+	// If we're in interactive mode, prompt for the project name
+	// for each stack that doesn't have one.
+	if cmdutil.Interactive() {
+		opts.ProjectsForDetachedStacks = cmd.projectsForDetachedStacks
+	}
+	return lb.Upgrade(ctx, &opts)
+}
+
+func (cmd *stateUpgradeCmd) projectsForDetachedStacks(stacks []tokens.Name) ([]tokens.Name, error) {
+	projects := make([]tokens.Name, len(stacks))
+	err := (&stateUpgradeProjectNameWidget{
+		Stdin:  cmd.Stdin,
+		Stdout: cmd.Stdout,
+		Stderr: cmd.Stderr,
+	}).Prompt(stacks, projects)
+	return projects, err
+}
+
+// stateUpgradeProjectNameWidget is a widget that prompts the user
+// for a project name for every stack that doesn't have one.
+//
+// It is used by the 'pulumi state upgrade' command
+// when it encounters stacks without a project name.
+type stateUpgradeProjectNameWidget struct {
+	Stdin  io.Reader // required
+	Stdout io.Writer // required
+	Stderr io.Writer // required
+}
+
+// Prompt prompts the user for a project name for each stack
+// and stores the result in the corresponding index of projects.
+//
+// The length of projects must be equal to the length of stacks.
+func (w *stateUpgradeProjectNameWidget) Prompt(stacks, projects []tokens.Name) error {
+	contract.Assertf(len(stacks) == len(projects),
+		"length of stacks (%d) must equal length of projects (%d)", len(stacks), len(projects))
+
+	if len(stacks) == 0 {
+		// Nothing to prompt for.
+		return nil
+	}
+
+	stdin, ok1 := w.Stdin.(terminal.FileReader)
+	stdout, ok2 := w.Stdout.(terminal.FileWriter)
+	if !ok1 || !ok2 {
+		// We're not using a real terminal, so we can't prompt.
+		// Pretend we're in non-interactive mode.
+		return nil
+	}
+
+	fmt.Fprintln(stdout, "Found stacks without a project name.")
+	fmt.Fprintln(stdout, "Please enter a project name for each stack, or enter to skip that stack.")
+	for i, stack := range stacks {
+		var project string
+		err := survey.AskOne(&survey.Input{
+			Message: fmt.Sprintf("Stack %s", stack),
+			Help:    "Enter a name for the project, or press enter to skip",
+		}, &project,
+			survey.WithStdio(stdin, stdout, w.Stderr),
+			survey.WithValidator(w.validateProject))
+		if err != nil {
+			return fmt.Errorf("prompt for %q: %w", stack, err)
+		}
+
+		projects[i] = tokens.Name(project)
+	}
+
+	return nil
+}
+
+func (w *stateUpgradeProjectNameWidget) validateProject(ans any) error {
+	proj, ok := ans.(string)
+	contract.Assertf(ok, "widget should have a string output, got %T", ans)
+
+	switch {
+	case proj == "": // empty string means skip
```

Comment:
Bit ick that we're rewriting this >100 IsName logic _again_, must be like 4 or 5 places now with this. At the least lets put a helper function somewhere that all these call sites can use.

Reply:
Yeah, I didn't like it either. I'll fix it.

Reply:
Done. Added to bottom of the stack https://github.com/pulumi/pulumi/pull/13165